File: src/main/java/com/upokecenter/mail/CollectionUtilities.java
package com.upokecenter.mail;
/*
Written by Peter O.
Any copyright to this work is released to the Public Domain.
In case this is not possible, this work is also
licensed under the Unlicense: https://unlicense.org/

 */

import java.util.*;

  final class CollectionUtilities {
private CollectionUtilities() {
}
    public static <TKey, TValue> boolean MapEquals(
      Map<TKey, TValue> mapA,
      Map<TKey, TValue> mapB) {
      if (mapA == null) {
        return mapB == null;
      }
      if (mapB == null) {
        return false;
      }
      if (mapA.size() != mapB.size()) {
        return false;
      }
      for (Map.Entry<TKey, TValue> kvp : mapA.entrySet()) {
        TValue valueB = null;
        boolean hasKey;
valueB = mapB.get(kvp.getKey());
hasKey=(valueB == null) ? mapB.containsKey(kvp.getKey()) : true;
        if (hasKey) {
          TValue valueA = kvp.getValue();
          if (!(((valueA) == null) ? ((valueB) == null) : (valueA).equals(valueB))) {
            return false;
          }
        } else {
          return false;
        }
      }
      return true;
    }

    public static <T> boolean ListEquals(List<T> listA, List<T> listB) {
      if (listA == null) {
        return listB == null;
      }
      if (listA.size() != listB.size()) {
        return false;
      }
      for (int i = 0; i < listA.size(); ++i) {
        if (!(((listA.get(i)) == null) ? ((listB.get(i)) == null) : (listA.get(i)).equals(listB.get(i)))) {
          return false;
        }
      }
      return true;
    }
  }

File: MailLib/PeterO/Mail/CollectionUtilities.cs
/*
Written by Peter O.
Any copyright to this work is released to the Public Domain.
In case this is not possible, this work is also
licensed under the Unlicense: https://unlicense.org/

 */
using System;
using System.Collections.Generic;

namespace PeterO.Mail {
  internal static class CollectionUtilities {
    public static bool MapEquals<TKey, TValue>(
      IDictionary<TKey, TValue> mapA,
      IDictionary<TKey, TValue> mapB) {
      if (mapA == null) {
        return mapB == null;
      }
      if (mapB == null) {
        return false;
      }
      if (mapA.Count != mapB.Count) {
        return false;
      }
      foreach (KeyValuePair<TKey, TValue> kvp in mapA) {
        TValue valueB = default(TValue);
        bool hasKey = mapB.TryGetValue(kvp.Key, out valueB);
        if (hasKey) {
          TValue valueA = kvp.Value;
          if (!Object.Equals(valueA, valueB)) {
            return false;
          }
        } else {
          return false;
        }
      }
      return true;
    }

    public static bool ListEquals<T>(IList<T> listA, IList<T> listB) {
      if (listA == null) {
        return listB == null;
      }
      if (listA.Count != listB.Count) {
        return false;
      }
      for (var i = 0; i < listA.Count; ++i) {
        if (!Object.Equals(listA[i], listB[i])) {
          return false;
        }
      }
      return true;
    }
  }
}
